src/google/protobuf/compiler/csharp/csharp_helpers.h

```diff
@@ -101,10 +101,13 @@ uint FixedMakeTag(const FieldDescriptor* descriptor);
 
 FieldGeneratorBase* CreateFieldGenerator(const FieldDescriptor* descriptor, int fieldOrdinal);
 
-bool HasRequiredFields(const Descriptor* descriptor);
-
-inline bool SupportFieldPresence(const FileDescriptor* file) {
-  return file->syntax() != FileDescriptor::SYNTAX_PROTO3;
+// Determines whether we're generating code for the proto representation of descriptors etc,
+// for use in the runtime. This is the only type which is allowed to use proto2 syntax,
+// and it generates internal classes.
+inline bool IsDescriptorProto(const FileDescriptor* descriptor) {
+  // TODO: Do this better! (Currently this depends on a hack in generate_protos.sh to rename
```

Comment:
Yes, that would be better. I still want to see whether there's something else we can do to remove that hack in the first place, too - depending on *exactly* what members we generate, we might be able to go back to it being Descriptor.

Now, would you like me to fix this TODO in this PR, or tackle it later? (I do have another set of changes mostly ready to submit as a PR once we've merged the current train in... JSON formatting, although a few changes are needed based on more recent information.)